### PR TITLE
Flushing queries

### PR DESCRIPTION
**Pull request by vilhena-stripe**
- Add `QueryCollector` and `NullFlusher`
- Remove `ignoreFlushes`

### Motivation
Handle queries separately

### Test plan
See included automated tests.

### CODE REVIEW

**Comment by jez** [comment]
looks like patrick requested jvilk; happy to look at it anyways just re-request me directly.

File: core/ErrorQueue.cc [hunk 1]
```diff
@@ -6,88 +6,31 @@
...
-} // namespace
-
 using namespace std;
 
 ErrorQueue::ErrorQueue(spdlog::logger &logger, spdlog::logger &tracer, shared_ptr<ErrorFlusher> errorFlusher)
     : errorFlusher(errorFlusher), owner(this_thread::get_id()), logger(logger), tracer(tracer){};
 
-pair<vector<unique_ptr<core::Error>>, vector<unique_ptr<core::lsp::QueryResponse>>>
-ErrorQueue::drainWithQueryResponses() {
+vector<unique_ptr<core::Error>> ErrorQueue::drainAllErrors() {
```
Comment (jvilk-stripe) [inline core/ErrorQueue.cc:14]: Can we remove this function?

File: main/lsp/QueryCollector.cc [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+} // namespace
+void QueryCollector::flushErrors(spdlog::logger &logger, const core::GlobalState &gs, core::FileRef file,
+                                 vector<unique_ptr<core::ErrorQueueMessage>> errors) {
+    for (auto &error : errors) {
+        if (error->kind == core::ErrorQueueMessage::Kind::QueryResponse) {
+            queryResponses.emplace_back(move(error->queryResponse));
+        }
+    }
+
+    stable_sort(queryResponses.begin(), queryResponses.end(), [](auto &left, auto &right) -> bool {
```
Comment (jvilk-stripe) [inline main/lsp/QueryCollector.cc:37]: This function is now going to sort query responses every time they are flushed, which could be expensive for "Find All References" requests spanning many files.

Can we move this into a function called `drainQueryResponses()` that returns a sorted vector? Then you can make `queryResponses` private.

File: main/lsp/test/lsp_preprocessor_test.cc [hunk 3]
```diff
@@ -115,8 +116,7 @@ class CountingTask final : public core::lsp::Task {
         preemptManager->assertTypecheckMutexHeld();
         // Emulate behavior of most LSP Tasks and drain all diagnostics and query responses.
         // This should never drain error queue items from the preempted task.
-        gs.errorQueue->drainWithQueryResponses();
-        CHECK(gs.errorQueue->ignoreFlushes);
+        gs.errorQueue->drainAllErrors();
```
Comment (jvilk-stripe) [inline main/lsp/test/lsp_preprocessor_test.cc:119]: I think this could be a `flushAllErrors` (so we can remove `drainAllErrors`)

File: main/pipeline/pipeline.cc [hunk 4]
```diff
@@ -1225,9 +1226,9 @@ namespace {
 core::FileHash computeFileHash(shared_ptr<core::File> forWhat, spdlog::logger &logger) {
     Timer timeit(logger, "computeFileHash");
     const static options::Options emptyOpts{};
-    unique_ptr<core::GlobalState> lgs = make_unique<core::GlobalState>((make_shared<core::ErrorQueue>(logger, logger)));
+    unique_ptr<core::GlobalState> lgs = make_unique<core::GlobalState>(
```
Comment (jvilk-stripe) [inline main/pipeline/pipeline.cc:1229]: 🎉

File: main/realmain.cc [hunk 5]
```diff
@@ -477,7 +478,8 @@ int realmain(int argc, char *argv[]) {
         logger->warn("LSP is disabled in sorbet-orig for faster builds");
         return 1;
 #else
-        gs->errorQueue->ignoreFlushes = true;
+        gs->errorQueue = make_shared<core::ErrorQueue>(gs->errorQueue->logger, gs->errorQueue->tracer,
```
Comment (jvilk-stripe) [inline main/realmain.cc:481]: You can just remove this; we'll replace it before doing any work with `gs`.

File: core/ErrorQueue.cc (hunk 1 above)
Comment (vilhena-stripe) [reply core/ErrorQueue.cc:14]: It's used in a number of tests

Comment (vilhena-stripe) [reply core/ErrorQueue.cc:14]: could remove it if `flushAllErrors` returned the vector instead

File: core/ErrorQueue.h [hunk 6]
```diff
@@ -15,32 +15,27 @@ namespace core {
 class ErrorQueue {
 private:
     void checkOwned();
-    UnorderedMap<core::FileRef, std::vector<std::unique_ptr<ErrorQueueMessage>>> drainAll();
     std::shared_ptr<ErrorFlusher> errorFlusher;
     const std::thread::id owner;
     UnorderedMap<core::FileRef, std::vector<std::unique_ptr<ErrorQueueMessage>>> collected;
     ConcurrentUnBoundedQueue<core::ErrorQueueMessage> queue;
+    UnorderedMap<core::FileRef, std::vector<std::unique_ptr<ErrorQueueMessage>>> drainAll();
```
Comment (jvilk-stripe) [inline core/ErrorQueue.h:22]: ? Why was this moved?

File: core/ErrorQueue.h [hunk 7]
```diff
@@ -15,32 +15,27 @@ namespace core {
...
 
     /** register a new error to be reported */
     void pushError(const GlobalState &gs, std::unique_ptr<Error> error);
-    void pushQueryResponse(std::unique_ptr<lsp::QueryResponse> response);
-    /** Extract all query responses. This discards all errors currently present in error Queue */
-    std::pair<std::vector<std::unique_ptr<core::Error>>, std::vector<std::unique_ptr<core::lsp::QueryResponse>>>
-    drainWithQueryResponses();
+    void pushQueryResponse(core::FileRef fromFile, std::unique_ptr<lsp::QueryResponse> response);
     bool isEmpty();
     /** Extract all errors. This discards all query responses currently present in error Queue */
```
Comment (jvilk-stripe) [inline core/ErrorQueue.h:38]: This comment should be removed.

File: test/pipeline_test_runner.cc [hunk 8]
```diff
@@ -134,14 +137,15 @@ class ExpectationHandler {
...
+        errorQueue->flushAllErrors(gs);
+        auto newErrors = errorCollector->drainErrors();
         errors.insert(errors.end(), make_move_iterator(newErrors.begin()), make_move_iterator(newErrors.end()));
     }
 
-    void clear() {
+    void clear(const core::GlobalState &gs) {
         got.clear();
-        errorQueue->drainAllErrors();
+        errorQueue->flushAllErrors(gs);
```
Comment (jvilk-stripe) [inline test/pipeline_test_runner.cc:148]: I think this also has to `drainErrors` on the `errorCollector` to throw the collected errors away.

File: test/pipeline_test_runner.cc [hunk 9]
```diff
@@ -421,9 +426,6 @@ TEST_CASE("PerPhaseTest") { // NOLINT
         ErrorAssertion::checkAll(test.sourceFileContents, RangeAssertion::getErrorAssertions(assertions), diagnostics);
     }
 
-    // Allow later phases to have errors that we didn't test for
-    errorQueue->drainAllErrors();
```
Comment (jvilk-stripe) [inline test/pipeline_test_runner.cc:425]: I think this still needs to `flushAllErrors` on the queue and `drainErrors` on the collector.

File: test/pipeline_test_runner.cc [hunk 10]
```diff
@@ -473,23 +475,23 @@ TEST_CASE("PerPhaseTest") { // NOLINT
...
-        handler.addObserved("resolve-tree", [&]() { return resolvedTree.tree->toString(*gs); });
-        handler.addObserved("resolve-tree-raw", [&]() { return resolvedTree.tree->showRaw(*gs); });
+        handler.addObserved(*gs, "resolve-tree", [&]() { return resolvedTree.tree->toString(*gs); });
+        handler.addObserved(*gs, "resolve-tree-raw", [&]() { return resolvedTree.tree->showRaw(*gs); });
     }
 
     handler.checkExpectations("[stress-incremental] ");
 
     // and drain all the remaining errors
-    errorQueue->drainAllErrors();
```
Comment (jvilk-stripe) [inline test/pipeline_test_runner.cc:492]: Ditto

File: core/ErrorQueue.h (hunk 6 above)
Comment (vilhena-stripe) [reply core/ErrorQueue.h:22]: ...user error?
